File: exe05.cc
#include "ns3/core-module.h"
#include "ns3/point-to-point-module.h"
#include "ns3/network-module.h"
#include "ns3/applications-module.h"
#include "ns3/wifi-module.h"
#include "ns3/mobility-module.h"
#include "ns3/csma-module.h"
#include "ns3/internet-module.h"
#include "ns3/propagation-loss-model.h"

#include <string>

using namespace ns3;

void ReceivePacket (Ptr<Socket> socket){
  NS_LOG_UNCOND ("Received one packet!");
}

NS_LOG_COMPONENT_DEFINE ("Simple-wifi-whit-csma-ping");

int main (int argc, char *argv[]){
  LogComponentEnable ("UdpClient", LOG_LEVEL_INFO);
  LogComponentEnable ("UdpServer", LOG_LEVEL_INFO);

  uint32_t nNodes = 2; //number of nodes
  int seed = 1; //simulation seed
  
  /*
   * Command line
   */
  CommandLine cmd;
  cmd.AddValue ("seed", "Seed", seed);
  cmd.AddValue ("nNodes", "Number of nodes", nNodes);
  cmd.Parse (argc, argv);
  
  /*
   * Set simulation seed
   */
  ns3::SeedManager::SetSeed (seed);
  
  /*
   * Create nodes:
   * There are one container with nNodes.
   */
  NodeContainer nodes;
  nodes.Create (nNodes);
  
  /*
   * Create Mobility Model:
   */
  MobilityHelper mobility;
  mobility.SetPositionAllocator ("ns3::GridPositionAllocator",
                                 "MinX", DoubleValue (0.0),
                                 "MinY", DoubleValue (0.0),
                                 "DeltaX", DoubleValue (10.0),
                                 "DeltaY", DoubleValue (10.0),
                                 "GridWidth", UintegerValue (2),
                                 "LayoutType", StringValue ("RowFirst"));
  mobility.Install (nodes);
  
  /*
   * Wifi:
   * The default state is defined as being an Adhoc MAC layer with an ARF rate 
   * control algorithm and both objects using their default attribute values. 
   * By default, configure MAC and PHY for 802.11a.
   */
  WifiHelper wifi = WifiHelper::Default ();
  wifi.SetRemoteStationManager ("ns3::ConstantRateWifiManager",
                                "MaxSsrc", UintegerValue(0), 
                                "MaxSlrc", UintegerValue(0), 
                                "RtsCtsThreshold", UintegerValue(0),
                                "DataMode", StringValue("OfdmRate54Mbps"), 
                                "ControlMode", StringValue("OfdmRate54Mbps"),
                                "NonUnicastMode",StringValue("OfdmRate54Mbps"));
  
  /* 
   * Create WifiChannel (comunication channel):
   * Create a channel helper in a default working state. By default, we create 
   * a channel model with a propagation delay equal to a constant, the speed of 
   * light, and a propagation loss based on a log distance model with a 
   * reference loss of 46.6777 dB at reference distance of 1m.
   * 
   * Create WifiPhy:
   * Create a phy helper without any parameter set. The user must set them all 
   * to be able to call Install later.
   */
  YansWifiChannelHelper wifiChannel = YansWifiChannelHelper::Default ();
  Ptr<YansWifiChannel> channel = wifiChannel.Create ();
  
  YansWifiPhyHelper wifiPhy = YansWifiPhyHelper::Default ();
  wifiPhy.SetChannel (channel);

  /*
   * Create WifiMac for each wifi node:
   * Create a mac helper in a default working state. i.e., this is an adhoc 
   * mac by default
   */
  NqosWifiMacHelper wifiMac;
  
  char ssid_str[100];
  sprintf (ssid_str, "%s", "myssid");
  std::cout << "ssid = " << ssid_str << std::endl;
  
  Ssid ssid = Ssid (ssid_str);
  wifiMac = NqosWifiMacHelper::Default ();
  wifiMac.SetType ("ns3::AdhocWifiMac", "Ssid", SsidValue (ssid));
  
  /*
   * Create netDevices:
   */
  NetDeviceContainer netDevices;
  netDevices = wifi.Install (wifiPhy, wifiMac, nodes);
  
  /*
   * Create stack protocols:
   */
  InternetStackHelper stack;
  stack.Install (nodes);

  /*
   * Create ip address:
   */
  Ipv4AddressHelper address;
  Ipv4InterfaceContainer interfaces;
  
  char ip_str[100];
  sprintf(ip_str, "%s", "192.168.1.0");
  std::cout << "ip = " << ip_str << std::endl;
  
  address.SetBase(ip_str, "255.255.255.0");
  interfaces = address.Assign (netDevices);

  /*
   * Traffic:
   */
  TypeId tid = TypeId::LookupByName ("ns3::UdpSocketFactory");
  Ptr<Socket> recvSink = Socket::CreateSocket(nodes.Get(0), tid);
  InetSocketAddress local = InetSocketAddress(Ipv4Address::GetAny(), 2001);
  recvSink->Bind (local);
  recvSink->SetRecvCallback(MakeCallback(&ReceivePacket));

  Ptr<Socket> source = Socket::CreateSocket (nodes.Get(1), tid);
  InetSocketAddress remote = InetSocketAddress(interfaces.GetAddress(0), 2001);
  source->Connect (remote);
  
  // connect all our nodes to a shared channel.
  NS_LOG_INFO ("Build Topology.");
  CsmaHelper csma;
  csma.SetChannelAttribute ("DataRate", DataRateValue (DataRate (5000000)));
  csma.SetChannelAttribute ("Delay", TimeValue (MilliSeconds (2)));
  csma.SetDeviceAttribute ("EncapsulationMode", StringValue ("Llc"));
  NetDeviceContainer devs = csma.Install (nodes);

   // assign ip addresses
  Address serverAddress;
  NS_LOG_INFO ("Assign ip addresses.");
  Ipv4InterfaceContainer addresses = address.Assign (devs);
  serverAddress = Address (addresses.GetAddress(1));
  //
// Create one udpServer applications on node one.
//
  uint16_t port = 4000;
  UdpServerHelper server (port);
  ApplicationContainer apps = server.Install (nodes.Get (1));
  apps.Start (Seconds (1.0));
  apps.Stop (Seconds (10.0));

//
// Create one UdpClient application to send UDP datagrams from node zero to
// node one.
//
  uint32_t MaxPacketSize = 1024;
  Time interPacketInterval = Seconds (0.05);
  uint32_t maxPacketCount = 30;
  UdpClientHelper client (serverAddress, port);
  client.SetAttribute ("MaxPackets", UintegerValue (maxPacketCount));
  client.SetAttribute ("Interval", TimeValue (interPacketInterval));
  client.SetAttribute ("PacketSize", UintegerValue (MaxPacketSize));
  apps = client.Install (nodes.Get (0));
  apps.Start (Seconds (2.0));
  apps.Stop (Seconds (10.0));

  Packet::EnablePrinting ();
  Simulator::Run ();
  Simulator::Destroy ();
  
  return 0;
}

File: exe05.py
import sys

import ns.applications
import ns.core
import ns.internet
import ns.mobility
import ns.network
import ns.wifi

def SetPosition(node, position):
	mobility = node.GetObject(ns.mobility.MobilityModel.GetTypeId())
	mobility.SetPosition(position)

def GetPosition(node):
	mobility = node.GetObject(ns.mobility.MobilityModel.GetTypeId())
	return mobility.GetPosition()

def AdvancePosition(node):
	pos = GetPosition(node);
	pos.x += 5.0
	if pos.x >= 210.0:
		return
	SetPosition(node, pos)
	ns.core.Simulator.Schedule(ns.core.Seconds(1.0), AdvancePosition, node)

def main(argv):

	ns.core.LogComponentEnable("UdpEchoClientApplication", ns.core.LOG_LEVEL_INFO)
	ns.core.LogComponentEnable("UdpEchoServerApplication", ns.core.LOG_LEVEL_INFO)

	# define a node container
	stations = ns.network.NodeContainer()
	
	# create 3 nodes
	stations.Create(3)

	# chose the element at 0 to be the AP of the wifi network
	ap = stations.Get(0)

	# contruct the wifi devices and the intercommunication channel between these wifi nodes 
	phy = ns.wifi.YansWifiPhyHelper.Default()
	channel = ns.wifi.YansWifiChannelHelper.Default()
	phy.SetChannel(channel.Create())

	# wifi helper
	wifi = ns.wifi.WifiHelper.Default()
	wifi.SetStandard(ns.wifi.WIFI_PHY_STANDARD_80211b)
	#wifi.SetRemoteStationManager("ns3::ArfWifiManager")
	wifi.SetRemoteStationManager(
		"ns3::ConstantRateWifiManager", 
		"DataMode", ns.core.StringValue("DsssRate5_5Mbps"))
	#	"DataMode", ns.core.StringValue("wifia-54mbs"))
	
	# SSID of the network
	ssid = ns.wifi.Ssid("wifi-home")

	# wifi Mac
	mac = ns.wifi.NqosWifiMacHelper.Default()
	mac.SetType("ns3::StaWifiMac",
                    "Ssid", ns.wifi.SsidValue(ssid),
                    "ActiveProbing", ns.core.BooleanValue(False))
    
	# network device container
	stationDevices = wifi.Install(phy, mac, stations)

	# configuring type for the AP
	mac.SetType("ns3::ApWifiMac",
                    "Ssid", ns.wifi.SsidValue(ssid),
                    "BeaconGeneration", ns.core.BooleanValue(True),
                    "BeaconInterval", ns.core.TimeValue(ns.core.Seconds(2.5)))
	apDevices = wifi.Install(phy, mac, ap)

	# mobility, I dont know that this does
	mobility = ns.mobility.MobilityHelper()
	mobility.Install(stations)
	ns.core.Simulator.Schedule(ns.core.Seconds(1.0), AdvancePosition, ap)

	#print help(ns.internet.InternetStackHelper)
	stackHelper = ns.internet.InternetStackHelper()
	stack = stackHelper.Install(stations)
	
	#Ipv4 address container
	address = ns.internet.Ipv4AddressHelper()
	address.SetBase(
		ns.network.Ipv4Address("10.1.1.0"),
		ns.network.Ipv4Mask("255.255.255.0"))
	stInterfaces = address.Assign(stationDevices)
	apInterfaces = address.Assign(apDevices)

	device = apDevices.Get(ap.GetId())

	echoServer = ns.applications.UdpEchoServerHelper(9)

	serverApps = echoServer.Install(stations.Get(1))
	serverApps.Start(ns.core.Seconds(1.0))
	serverApps.Stop(ns.core.Seconds(10.0))

	echoClient = ns.applications.UdpEchoClientHelper(stInterfaces.GetAddress(1), 9)
	echoClient.SetAttribute("MaxPackets", ns.core.UintegerValue(1))
	echoClient.SetAttribute("Interval", ns.core.TimeValue(ns.core.Seconds(1.0)))
	echoClient.SetAttribute("PacketSize", ns.core.UintegerValue(1024))

	clientApps = echoClient.Install(stations.Get(2))
	clientApps.Start(ns.core.Seconds(1.0))
	clientApps.Stop(ns.core.Seconds(2.0))

	ns.core.Simulator.Run()
	ns.core.Simulator.Destroy()

	return 0

if __name__ == '__main__':
	sys.exit(main(sys.argv))